tools/dds/dds-server/lrs-device-controller.cpp

```diff
@@ -2,99 +2,205 @@
+                                                 requested_stream_name,
+                                                 dds_profile );
+        rs_profiles_to_open.push_back( rs2_profile );
+
+        auto dds_vp = std::dynamic_pointer_cast< dds_video_stream_profile >( dds_profile );
+        realdds::image_header header;
+        header.format = dds_profile->format().to_rs2();
+        header.width = dds_vp ? dds_vp->width() : 0;
+        header.height = dds_vp ? dds_vp->height() : 0;
+        realdds_streams_to_start.push_back( std::make_pair( requested_stream_name, std::move( header ) ) );
     }
+
+    //Start streaming
+    //TODO - currently assumes all streams are from one sensor. Add support for multiple sensors
+    _dds_device_server->start_streaming( realdds_streams_to_start );
+    _sensors[sensor_index].open( rs_profiles_to_open );
+    _sensors[sensor_index].start( [&]( rs2::frame f ) {
+        _dds_device_server->publish_image( f.get_profile().stream_name(), static_cast< const uint8_t * >( f.get_data() ), f.get_data_size() );
+    } );
+    std::cout << realdds_streams_to_start[0].first << " stream started" << std::endl;
 }
 
-void lrs_device_controller::stop_stream( rs2_stream stream )
+void lrs_device_controller::stop_streaming( const json & msg )
 {
-    auto sensor_to_erase_it = stream_to_rs2_sensor.find( stream );
-    if( sensor_to_erase_it == stream_to_rs2_sensor.end() )
+    auto stream_names = msg["stream-names"];
+
+    size_t sensor_index = 0;
+    for ( std::string requested_stream_name : stream_names )
     {
-        throw std::runtime_error( "Cannot stop stream:"
-                                  + std::string( rs2_stream_to_string( stream ) ) + " as it is not streaming" );
+        //Find sensor to close
+        if ( find_sensor( requested_stream_name, sensor_index ) )
+            break; //TODO - currently assumes all streams are from one sensor. Add support for multiple sensors
+
+        if ( sensor_index == _sensors.size() )
+            throw std::runtime_error( "Could not find sensor to close `" + requested_stream_name + "` for" );
     }
 
-    stream_to_rs2_sensor.erase( sensor_to_erase_it );
+    //Stop streaming
+    _sensors[sensor_index].stop();
+    _sensors[sensor_index].close();
+    _dds_device_server->stop_streaming( stream_names );
+    std::cout << _sensors[sensor_index].get_info( RS2_CAMERA_INFO_NAME ) << " closed. Streams requested to stop: " << stream_names << std::endl;
 }
 
-void lrs_device_controller::stop_all_streams()
+bool tools::lrs_device_controller::find_sensor( const std::string & requested_stream_name, size_t & sensor_index )
 {
-    stream_to_rs2_sensor.clear();
+    //Iterate over all the sensors, if a sensor have a profile with a name
+    //corresponding to the requested, we have found our sensor.
+    for ( sensor_index = 0; sensor_index < _sensors.size(); ++sensor_index )
+    {
+        for ( auto & profile : _sensors[sensor_index].get_stream_profiles() )
```

Comment:
Why not do a `sensor.get_info( RS2_CAMERA_INFO_NAME )` to get its name?
We could do this once and store a map of sensor-to-index, even in the ctor...

Reply:
I forget how confusing it is...
You're looking for a stream name, not a sensor name.
Still, I think we need to be extra careful here: stopping a stream is not the same as stopping a sensor. I.e., you can close the IR streams but leave Depth running. I'd even make that a unit-test.

Reply:
You are right that stopping a stream should not, theoretically, stop other streams, but this is not the current use case.
This tool is giving LibRS users the behavior that they expect from a USB camera, and the LibRS API does not enable to stop just a stream, but a whole sensor.
In the future we may decide to change this behavior, but not for this PR.
(Plus it would really change if we'll open/close streams based on readers availability)

Reply:
I'm not saying you should take care of this "use-case" in this specific tool. But the API and the triggers for everything need to be understood and acted on now, not later.

The way we open/close streams is in realdds and the camera-host communication, not the tool. Here we need to envision all use-cases and make sure the API is suitable for multiple readers *if needed*.

Don't be limited by the use-case.